WalletWasabi.Fluent/ViewModels/Wallets/Buy/Workflows/Workflow.cs

```diff
@@ -0,0 +1,127 @@
+using System.Reactive.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using ReactiveUI;
+using WalletWasabi.BuyAnything;
+using WalletWasabi.Fluent.ViewModels.Wallets.Buy.Workflows.ShopinBit;
+using WalletWasabi.Wallets;
+using WalletWasabi.Extensions;
+using WalletWasabi.Logging;
+
+namespace WalletWasabi.Fluent.ViewModels.Wallets.Buy.Workflows;
+
+public abstract partial class Workflow : ReactiveObject
+{
+	[AutoNotify] private IWorkflowStep? _currentStep;
+	[AutoNotify] private Conversation _conversation;
+	[AutoNotify] private bool _isCompleted;
+	[AutoNotify] private bool _isDeletedInSib;
+
+	protected Workflow(Conversation conversation)
+	{
+		_conversation = conversation;
+
+		BindCurrentStepConversation();
+
+		this.WhenAnyValue(x => x.Conversation)
+			.Do(x =>
+			{
+				if (CurrentStep is { } step)
+				{
+					step.Conversation = x;
+				}
+			})
+			.Subscribe();
+	}
+
+	public event EventHandler<Exception>? OnStepError;
+
+	public abstract IMessageEditor MessageEditor { get; }
+
+	public abstract Task ExecuteAsync(CancellationToken token);
+
+	protected async Task ExecuteStepAsync(IWorkflowStep step)
+	{
+		CurrentStep = step;
+		step.Conversation = Conversation;
+
+		// this is looped until Step execution is successfully completed.
+		// If it errors out, then the Workflow won't move forward to the next step.
+		// All Steps should be be able to be re-executed more than once, gracefully.
+		while (true)
```

Comment:
If an error happens, this will pop up a dialog infinitely. Even if you close the Bab dialog.